### PR TITLE
Refactor Parser and YamlReader

### PR DESCRIPTION
**Pull request by ramchaik**
After looking through the codebase, I was impressed by the code coverage and thought it was pretty amazing.

I made the decision to refactor certain small portions because I wanted to contribute to the project.

This PR refactors some methods of the Parser and YamlReader classes.

##### Decompose type conditional for implicit document in `initProductionTable`

- The `produce()` method calls the helper method `isNextTokenImplicitDocument()` to check if the next token indicates an implicit document. This makes the condition clearer and easier to understand.

- The `isNextTokenImplicitDocument()` method encapsulates the condition logic, making it self-explanatory.

- If the condition is met, the `parseImplicitDocument()` method is called, which contains the logic for adding items to the parse stack. This separates concerns and improves code readability.

##### Rename method `valueConvertedNumber()` to `parseNumber()`
- The method valueConvertedNumber has been renamed to parseNumber for improved clarity and consistency with naming conventions.

- Enhances the readability of the codebase and better reflects the purpose of the method, which is to parse a string value into a Number object.

##### Extract method `readScalarValue()` from `readValueInternal()`
- Extract the deserialization of scalar values into a new method.
- Increases modularity of readValueInternal.
- Improved the documentation by adding Javadoc for the `readScalarValue()` method.

---
### Checklist
Checklist I followed before raising the PR.
- [x] Build the project
- [x] Run the tests after the refactor

Screenshot 2024-03-16 at 2 06 48 PM

### CODE REVIEW

File: src/com/esotericsoftware/yamlbeans/YamlReader.java [hunk 1]
```diff
@@ -49,7 +49,7 @@
 public class YamlReader implements AutoCloseable {
 	private final YamlConfig config;
 	Parser parser;
-	private final Map<String, Object> anchors = new HashMap();
+	private final Map<String, Object> anchorMap = new HashMap();
```
Comment (NathanSweet) [inline src/com/esotericsoftware/yamlbeans/YamlReader.java:52]: There's no reason to put the type in the name. I don't think this needs to be renamed.

**Comment by NathanSweet** [comment]
Just the one comment. The other changes look good!

File: src/com/esotericsoftware/yamlbeans/YamlReader.java (hunk 1 above)
Comment (ramchaik) [reply src/com/esotericsoftware/yamlbeans/YamlReader.java:52]: sure, I will remove this.

Comment (ramchaik) [reply src/com/esotericsoftware/yamlbeans/YamlReader.java]: I have removed this change in my recent commit (9af34d4582a71e8d8396fe805198a1952354f073)

**Comment by NathanSweet** [comment]
Cheers!

**Comment by ramchaik** [comment]
Thanks @NathanSweet!

**Pull request merged**